piecestore/piecestore.go

```diff
@@ -0,0 +1,222 @@
+package piecestore
+
+import (
+	"fmt"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/google/uuid"
+	"github.com/hashicorp/go-multierror"
+	"github.com/ipfs/go-cid"
+	bstore "github.com/ipfs/go-ipfs-blockstore"
+	"github.com/ipld/go-car/v2"
+	carv2 "github.com/ipld/go-car/v2"
+	"github.com/ipld/go-car/v2/blockstore"
+	carindex "github.com/ipld/go-car/v2/index"
+	mh "github.com/multiformats/go-multihash"
+	"io"
+)
+
+type SectionReader interface {
+	io.Reader
+	io.ReaderAt
+	io.Seeker
+}
+
+type Sealer interface {
+	// GetReader returns a reader over a piece. If there is no unsealed copy, returns ErrSealed.
+	GetReader(id abi.SectorNumber, offset abi.PaddedPieceSize, length abi.PaddedPieceSize) (SectionReader, error)
+}
+
+type CarIndex interface {
+	Add(pieceCid cid.Cid, itidx carindex.IterableIndex) error
+	IsIndexed(pieceCid cid.Cid) bool
+	Index(pieceCid cid.Cid) (carindex.IterableIndex, error)
+	Delete(pieceCid cid.Cid) error
+}
+
+type MHToPieceIndex interface {
+	Add(pieceCid cid.Cid, itidx carindex.IterableIndex) error
+	IsIndexed(pieceCid cid.Cid) bool
+	PiecesContaining(m mh.Multihash) ([]cid.Cid, error)
+	Delete(pieceCid cid.Cid) error
+}
+
+type DealStore interface {
+	Add(pieceCid cid.Cid, dealInfo DealInfo) error
+	List(pieceCid cid.Cid) ([]DealInfo, error)
+	Delete(pieceCid cid.Cid, dealUuid uuid.UUID) (bool, error)
+}
+
+type PieceStore struct {
+	dealStore      DealStore
+	carIndex       CarIndex
+	mhToPieceIndex MHToPieceIndex
+	sealer         Sealer
+}
+
+// DealInfo is information about a single deal for a given piece
+type DealInfo struct {
+	DealUuid     uuid.UUID
+	PublishCid   cid.Cid
+	SectorID     abi.SectorNumber
+	Offset       abi.PaddedPieceSize
+	Length       abi.PaddedPieceSize
+	PaddedLength abi.PaddedPieceSize
+}
+
+func (ps *PieceStore) GetPieceDeals(pieceCid cid.Cid) ([]DealInfo, error) {
+	deals, err := ps.dealStore.List(pieceCid)
+	if err != nil {
+		return nil, fmt.Errorf("listing deals for piece %s: %w", pieceCid, err)
+	}
+
+	return deals, nil
+}
+
+func (ps *PieceStore) AddDealForPiece(pieceCid cid.Cid, dealInfo DealInfo) error {
+	// Add deal to list of deals for this piece
+	if err := ps.dealStore.Add(pieceCid, dealInfo); err != nil {
+		return fmt.Errorf("saving deal %s to store: %w", dealInfo.DealUuid, err)
+	}
+
+	// Perform indexing of piece
+	if err := ps.addIndexForPiece(pieceCid); err != nil {
+		return fmt.Errorf("adding index for piece %s: %w", pieceCid, err)
+	}
+
+	return nil
+}
+
+func (ps *PieceStore) addIndexForPiece(pieceCid cid.Cid) error {
+	// Check if the indexes have already been added
+	if ps.carIndex.IsIndexed(pieceCid) && ps.mhToPieceIndex.IsIndexed(pieceCid) {
+		return nil
+	}
+
```

Comment:
@dirkmc maybe we should add debug timers here, so that we have an idea how long it takes to fetch the piece and build an index.

Reply:
Yeah good idea to add some kind of tracing